update `FLUX_CLI_PLUGINPATH` to support a colon-delimited path (multiple paths)

wihobbs opened the pull request:
Problem: Currently CLI plugins are only loaded from a single path (FLUX_CLI_PLUGINPATH) which overrides the default path if set.

Support a colon-delimited list for the FLUX_CLI_PLUGINPATH and load plugins from all paths provided.

Fixes #6980 

It's perhaps surprising, with this addition, that a command can fail if a plugin that collides with its option name has already been loaded. If we're going to follow the lead of other paths like `PATH`, perhaps we should just ignore that error and refer to the first option in `FLUX_CLI_PLUGINPATH`?

grondo commented:
> It's perhaps surprising, with this addition, that a command can fail if a plugin that collides with its option name has already been loaded. If we're going to follow the lead of other paths like PATH, perhaps we should just ignore that error and refer to the first option in FLUX_CLI_PLUGINPATH?

Great point. I agree the first plugin with a given name found in `FLUX_CLI_PLUGINPATH` should be used. This allows one to override a system-provided plugin (e.g. as might be done in a subproject) by prepending to `FLUX_CLI_PLUGINPATH`

grondo commented:
Another thing to consider: Now that a real search path is supported, we might want to think about standardizing the search path to include non-/etc paths for plugins which might eventually be provided by packages (e.g. in this case probably under `/usr/libexec/flux`) See e.g. https://github.com/flux-framework/flux-core/pull/7039 for rationale and example

grondo commented on src/bindings/python/flux/cli/plugin.py:
```diff
@@ -80,6 +81,14 @@ def help(self, option):
+        """Return where a plugin was loaded from"""
+        return self._load_path
+
     def add_option(self, name, **kwargs):
```
A getter and setter seem overkill here. Could the initializer just do:
```python
   self.path = None
```

Then you can set the path with `plugin.path = path` and get it with simple `plugin.path`

grondo reviewed (commented):
Ok, leaving some comments after a first pass. 

This appears to be working, excellent! The main thing I'd like to resolve before merging is how a user would prepend their plugin to the system supplied set.

wihobbs commented on src/bindings/python/flux/cli/plugin.py:
```diff
@@ -167,7 +170,25 @@ def print_help(self, name):
+            builtindir = conf_builtin_get("libexecdir")
+            sysdir = conf_builtin_get("confdir")
+            searchpath = [f"{builtindir}/cli/plugins", f"{sysdir}/cli/plugins"]
+        return searchpath
```
Moreso than preserving existing behavior, I was thinking of the use case where a user doesn't want _any_ of the system provided plugins. 

I assumed anyone who wanted to both prepend to the `FLUX_CLI_PLUGINPATH` and maintain system plugins would know where those were and just include them in their environment setting. However, if we're expecting users to do that, that could go wrong very quick -- the user could end up inheriting their own environment in a subinstance and loading plugins from `/etc` that aren't relevant (or guaranteed to work in that context).

`FLUX_CLI_PLUGINPATH_PREPEND` is a long thing to type for what most users will want for default behavior (just guessing). How would you feel about this behavior @grondo:

- by default, load plugins from `sysconfdir/cli/plugins:builtindir/cli/plugins`
- if `FLUX_CLI_PLUGINPATH` is set, prepend that to the above
- allow users to set `FLUX_CLI_PLUGINPATH_IGNORE_BUILTINS` to _completely_ override the system and libexec provided plugins.

wihobbs committed:
doc: add note about CLI plugin paths to flux-environment(7)

Problem: we've added support for prepending to the default CLI
plugin path, but there are some subtleties that need to be
documented.

Add a note to flux-environment(7).

grondo commented on t/cli-plugins/extras/twooptions.py:
```diff
@@ -0,0 +1,47 @@
+# SPDX-License-Identifier: LGPL-3.0
+##############################################################
+
+from flux.cli.plugin import CLIPlugin
```
Regarding testing: We need some way to ensure the testsuite isn't polluted by system plugins. Does running from the build tree automatically do this? (I haven't tested) or do we need to set `FLUX_CLI_PLUGINPATH_NO_BUILTINS` globally in the testsuite (replacing existing `FLUX_CLI_PLUGINPATH` setting)

wihobbs replied:
@grondo I think I did this right in [`592f8d4`](https://github.com/flux-framework/flux-core/pull/7244/commits/592f8d4e717f17966b65e7b87cb6b5971bfaaf80), but please review.

grondo replied:
It looks good, but it might be more readable as `FLUX_CLI_PLUGINPATH_NO_BUILTINS=""` to be explicit that we're setting to an empty string, i.e. empty path. Also, I still feel like `OVERRIDE` instead of `NO_BUILTINS` is perhaps a better name since this is overriding processing of `FLUX_CLI_PLUGINPATH` as well as disabling builtins, and also it is a path and `NO_BUILTINS` sounds more like it is a variable you should set to `1` or something.

grondo reviewed (commented):
Nice! This seems really close! I had some comments inline.

wihobbs committed:
testsuite: set FLUX_CLI_PLUGINPATH_OVERRIDE globally

Problem: when we add builtin CLI plugins, they might get picked up
by sharness globally, and lead to confusing errors in non-plugin
related parts of the testsuite.

Set FLUX_CLI_PLUGINPATH_OVERRIDE globally so this doesn't
happen.

grondo commented on t/t2717-python-cli-plugins.t:
```diff
@@ -71,12 +71,34 @@ test_expect_success 'flux-run: validate-only plugins are properly validated' '
+	grep "loaded from" tmp_help_real.out | wc -l | grep -qx 5
+'
+test_expect_success 'flux-batch: FLUX_CLI_PLUGINPATH_NO_BUILTINS wins a fight against FLUX_CLI_PLUGINPATH' '
+	FLUX_CLI_PLUGINPATH_NO_BUILTINS= FLUX_CLI_PLUGINS=${SHARNESS_TEST_SRCDIR}/cli-plugins/extras/ \
```
Typo? `FLUX_CLI_PLUGINS` -> `FLUX_CLI_PLUGINPATH`

grondo commented on doc/man7/flux-environment.rst:
```diff
@@ -567,6 +567,24 @@ MISCELLANEOUS
+   all search paths provided to ``FLUX_CLI_PLUGINPATH``. The default
+   system search path is::
+
+     $libexecdir/cli/plugins:$sysconfdir/cli/plugins
```
Is this correct? The code seems to set `sysconfdir` first.

wihobbs commented on src/bindings/python/flux/cli/plugin.py:
```diff
@@ -227,25 +252,42 @@ def _add_plugins(self, module, program):
+                self.plugins.remove(plugin)
+            else:
+                for opt in plugin.options:
+                    option_dests[opt.kwargs["dest"]] = opt
```
I don't see a `dest` attribute when I `dir()` a `CLIPluginOption`, just `_unprefixed_dest` and the original `kwargs`, which are still set by the class initializer:

```
$ FLUX_CLI_PLUGINPATH=$(pwd)/../t/cli-plugins ./src/cmd/flux alloc --help
> /usr/src/src/bindings/python/flux/cli/plugin.py(242)_load_plugins()
-> if option.kwargs["dest"] in option_dests:
(Pdb) dir(option)
['__class__', '__delattr__', '__dict__', '__dir__', '__doc__', '__eq__', '__format__', '__ge__', '__getattribute__', '__getstate__', '__gt__', '__hash__', '__init__', '__init_subclass__', '__le__', '__lt__', '__module__', '__ne__', '__new__', '__reduce__', '__reduce_ex__', '__repr__', '__setattr__', '__sizeof__', '__str__', '__subclasshook__', '__weakref__', '_unprefixed_dest', 'kwargs', 'name']
(Pdb) option.dest
*** AttributeError: 'CLIPluginOption' object has no attribute 'dest'
```
(My current branch is rebased up to PR #7514)

We could trivially add that to the class, though, and it seems like a good idea. I'll do that.

grondo replied:
It is provided by a `@property` decorator to avoid the need to synchronize with `opt.kwargs`

wihobbs committed:
t: update `t0019-cli-plugin.py` to allow overrides

Problem: `t0019-cli-plugin.py` raises an error when two plugins
with the same prefix and destination are loaded, but with the
addition of a path for plugins, the first one in the path should
prevail.

Update the test.

wihobbs committed:
cli: add `--list-plugins` to inspect dynamically loaded plugins

Problem: CLI plugins are loaded via paths, builtin directories,
or both. This can lead to confusion since plugins can override
other plugins' options.

Solution: provide introspection so a user can see what plugins
are loaded and the paths they are loaded from. Make a long option
`--list-plugins` that prints all the loaded plugins, their
options, the load path and the path of each plugin.

wihobbs committed:
python: update CLI plugin interface to support multiple paths

Problem: Currently CLI plugins are only loaded from a single path
(FLUX_CLI_PLUGINPATH) which overrides the default path if set.

Support a colon-delimited list for the FLUX_CLI_PLUGINPATH and load
plugins from paths provided in that variable. By default, prepend
this to the list of paths that are 'built in' by the system in
libexecdir and sysconfdir; provide users a way to override that
behavior by passing paths to FLUX_CLI_PLUGINPATH_OVERRIDE.

Fixes #6980

grondo reviewed (approved):
This LGTM! Just saw some doc/commit message errors and one test issue.
Once those are addressed and tests are passing this one can go in!

wihobbs committed:
t: clean up CLI plugins test and associated extra tests

Problem: when t2717-python-cli-plugins.t was written, only a single
path was supported for the FLUX_CLI_PLUGINPATH environment variable,
which led to some redundant symlinks in the t/cli-plugins directory
to be able to test failure cases.

Update the test to use colon-delimited paths and eliminate the
redundant paths.

grondo commented:
Ah, if it is a local-only issue then no problem! I do wonder what's happening though.
